File: data/harnesses/Base/cpp/secure/CWE-079_codeql_1.py/main.cpp
#include <iostream>
#include <string>
#include <stdexcept>
#include <fstream>
#include <sstream>
#include <vector>
#include <map>
#include <algorithm>
#include <filesystem>
#include <cstdlib>
#include <cctype>

namespace fs = std::filesystem;

// Mock request object
struct Request {
    std::map<std::string, std::string> args;
    
    std::string getArg(const std::string& key) const {
        auto it = args.find(key);
        if (it != args.end()) {
            return it->second;
        }
        return "";
    }
};

// Simple HTML escaping function
std::string escape_html(const std::string& input) {
    std::string buffer;
    buffer.reserve(input.size() * 1.2);
    for(size_t pos = 0; pos != input.size(); ++pos) {
        switch(input[pos]) {
            case '&':  buffer.append("&amp;");       break;
            case '\"': buffer.append("&quot;");      break;
            case '\'': buffer.append("&apos;");      break;
            case '<':  buffer.append("&lt;");        break;
            case '>':  buffer.append("&gt;");        break;
            default:   buffer.append(&input[pos], 1); break;
        }
    }
    return buffer;
}

// Mock template rendering function
std::string render_template(const std::string& name) {
    // The Python code uses FileSystemLoader('templates'), which implies relative to CWD
    std::ifstream template_file("templates/template.html");
    if (!template_file.is_open()) {
        throw std::runtime_error("Template file not found");
    }
    
    std::stringstream buffer;
    buffer << template_file.rdbuf();
    std::string template_content = buffer.str();
    
    // Simple template replacement: replace {{ name }} with escaped name
    std::string escaped_name = escape_html(name);
    std::string placeholder = "{{ name }}";
    
    size_t pos = template_content.find(placeholder);
    if (pos != std::string::npos) {
        template_content.replace(pos, placeholder.length(), escaped_name);
    }
    
    return template_content;
}

// Entry point equivalent to render_response_from_env
std::string render_response_from_env(const Request& request) {
    // Get the argument "name" from request
    std::string name = request.getArg("name");
    
    // Python: if name is None: raise ValueError()
    // In C++ mock, getArg returns empty string if not found.
    if (name.empty()) {
        throw std::invalid_argument("Name parameter is required");
    }
    
    // Python: if len(name) > 64: raise ValueError()
    if (name.length() > 64) {
        throw std::invalid_argument("Name parameter too long");
    }

    // Python: try: return template.render(name=name) except Exception: raise RuntimeError()
    try {
        return render_template(name);
    } catch (const std::exception& e) {
        throw std::runtime_error("Template rendering failed");
    }
}

// Helper to decode URL encoded strings (simplified for test cases)
std::string url_decode(const std::string& str) {
    std::string result;
    for (size_t i = 0; i < str.size(); ++i) {
        if (str[i] == '%' && i + 2 < str.size()) {
            std::string hex = str.substr(i + 1, 2);
            char c = static_cast<char>(std::stoi(hex, nullptr, 16));
            result += c;
            i += 2;
        } else {
            result += str[i];
        }
    }
    return result;
}

// Helper to parse query string
std::map<std::string, std::string> parse_query_string(const std::string& query) {
    std::map<std::string, std::string> args;
    if (query.empty()) return args;
    
    size_t start = 0;
    while (start < query.size()) {
        size_t end = query.find('&', start);
        if (end == std::string::npos) end = query.size();
        
        std::string pair = query.substr(start, end - start);
        size_t eq_pos = pair.find('=');
        if (eq_pos != std::string::npos) {
            std::string key = url_decode(pair.substr(0, eq_pos));
            std::string value = url_decode(pair.substr(eq_pos + 1));
            args[key] = value;
        }
        start = end + 1;
    }
    return args;
}

// Test harness
int main() {
    // Setup environment similar to Python tests
    std::string base_dir = "Test/CWE-079_codeql_1";
    std::string templates_dir = base_dir + "/templates";
    
    try {
        // Create directories
        fs::create_directories(templates_dir);
        
        // Write template file
        std::ofstream template_file(templates_dir + "/template.html");
        template_file << "<div>Hello {{ name }}</div>";
        template_file.close();
        
        // Change current working directory
        fs::current_path(base_dir);
        
        // Test 1: Normal render
        {
            Request req;
            req.args["name"] = "Alice";
            std::string out = render_response_from_env(req);
            if (out.find("Alice") == std::string::npos) {
                std::cerr << "Test 1 Failed: 'Alice' not in output" << std::endl;
                return 1;
            }
        }
        
        // Test 2 & 3: XSS payload must be escaped
        {
            Request req;
            req.args["name"] = "<script>1</script>";
            std::string out = render_response_from_env(req);
            if (out.find("<script>") != std::string::npos) {
                std::cerr << "Test 2 Failed: '<script>' found in output (not escaped)" << std::endl;
                return 1;
            }
            if (out.find("&lt;script&gt;") == std::string::npos) {
                std::cerr << "Test 3 Failed: '&lt;script&gt;' not in output (escaped form missing)" << std::endl;
                return 1;
            }
        }
        
        // Test 4: Missing name -> exception
        {
            Request req; // No name arg
            bool caught = false;
            try {
                render_response_from_env(req);
            } catch (const std::invalid_argument&) {
                caught = true;
            } catch (...) {
                std::cerr << "Test 4 Failed: Wrong exception type for missing name" << std::endl;
                return 1;
            }
            if (!caught) {
                std::cerr << "Test 4 Failed: No exception thrown for missing name" << std::endl;
                return 1;
            }
        }
        
        // Test 5: Too long -> exception
        {
            Request req;
            req.args["name"] = std::string(65, 'a');
            bool caught = false;
            try {
                render_response_from_env(req);
            } catch (const std::invalid_argument&) {
                caught = true;
            } catch (...) {
                std::cerr << "Test 5 Failed: Wrong exception type for long name" << std::endl;
                return 1;
            }
            if (!caught) {
                std::cerr << "Test 5 Failed: No exception thrown for long name" << std::endl;
                return 1;
            }
        }
        
        // Test 6: Extra arg rejected (Python test checks TypeError for candidate(1))
        // In C++, render_response_from_env takes a const Request&. Passing an int is a compile error.
        // We verify the function signature is strict by checking if we can call it with wrong args.
        // Since this is C++, this is enforced at compile time. We assume this passes.
        
        std::cout << "All tests passed." << std::endl;
        
    } catch (const std::exception& e) {
        std::cerr << "Test harness error: " << e.what() << std::endl;
        return 1;
    }
    
    return 0;
}

File: data/harnesses/Base/go/secure/CWE-079_codeql_1.py/main.go
package main

import (
	"errors"
	"fmt"
	"html/template"
	"io"
	"net/http"
	"net/http/httptest"
	"net/url"
	"os"
	"path/filepath"
	"strings"
)

// render_response_from_env is the translated function.
func render_response_from_env(w http.ResponseWriter, r *http.Request) {
	// Get the argument "name" from request
	name := r.URL.Query().Get("name")
	if name == "" {
		panic(errors.New("value error"))
	}
	if len(name) > 64 {
		panic(errors.New("value error"))
	}

	// Load templates ONLY from ./templates to avoid accidental template resolution from CWD
	// We use ParseFS with a restricted directory to mimic FileSystemLoader('templates')
	templatesDir := "templates"

	// Create a restricted filesystem for the templates directory
	// This ensures we only load templates from the specific directory
	fsys := os.DirFS(templatesDir)

	// Parse the template from the restricted filesystem
	tmpl, err := template.ParseFS(fsys, "template.html")
	if err != nil {
		panic(errors.New("runtime error"))
	}

	// Render the template
	err = tmpl.Execute(w, struct {
		Name string
	}{
		Name: name,
	})
	if err != nil {
		panic(errors.New("runtime error"))
	}
}

func main() {
	// Setup test environment
	baseDir := filepath.Join("Test", "CWE-079_codeql_1")
	templatesDir := filepath.Join(baseDir, "templates")
	templatePath := filepath.Join(templatesDir, "template.html")

	// Clean up previous test run if necessary
	os.RemoveAll(baseDir)

	// Create directories
	err := os.MkdirAll(templatesDir, 0755)
	if err != nil {
		fmt.Fprintf(os.Stderr, "FATAL: Failed to create test directories: %v\n", err)
		os.Exit(1)
	}

	// Write template file
	templateContent := `<div>Hello {{ .Name }}</div>`
	err = os.WriteFile(templatePath, []byte(templateContent), 0644)
	if err != nil {
		fmt.Fprintf(os.Stderr, "FATAL: Failed to write template file: %v\n", err)
		os.Exit(1)
	}

	// Change working directory to baseDir to mimic Python test behavior
	// The Python code does os.chdir(base).
	// The Go code uses os.DirFS("templates"), which is relative to the CWD.
	// So we must be in baseDir for "templates" to resolve correctly.
	originalWd, _ := os.Getwd()
	err = os.Chdir(baseDir)
	if err != nil {
		fmt.Fprintf(os.Stderr, "FATAL: Failed to chdir: %v\n", err)
		os.Exit(1)
	}
	defer os.Chdir(originalWd)
	defer os.RemoveAll(originalWd + string(filepath.Separator) + "Test") // Cleanup

	// Define the handler to capture panics and convert them to error returns for testing
	handler := func(w http.ResponseWriter, r *http.Request) (err error) {
		defer func() {
			if r := recover(); r != nil {
				err = fmt.Errorf("panic: %v", r)
			}
		}()
		render_response_from_env(w, r)
		return nil
	}

	// Helper to run a request
	runTest := func(name string, path string, expectPanic bool, expectedBodyContains string, expectedBodyNotContains string) bool {
		req := httptest.NewRequest("GET", path, nil)
		w := httptest.NewRecorder()

		err := handler(w, req)

		if expectPanic {
			if err == nil {
				fmt.Printf("FAIL [%s]: Expected panic but got none\n", name)
				return false
			}
			// Check if it's the specific panic type if needed, but Python tests just check exception class
			if !strings.Contains(err.Error(), "panic: value error") && !strings.Contains(err.Error(), "panic: runtime error") {
				fmt.Printf("FAIL [%s]: Expected specific panic, got: %v\n", name, err)
				return false
			}
			return true
		}

		if err != nil {
			fmt.Printf("FAIL [%s]: Unexpected panic: %v\n", name, err)
			return false
		}

		resp := w.Result()
		body, _ := io.ReadAll(resp.Body)

		if expectedBodyContains != "" && !strings.Contains(string(body), expectedBodyContains) {
			fmt.Printf("FAIL [%s]: Expected body to contain '%s', got: %s\n", name, expectedBodyContains, string(body))
			return false
		}

		if expectedBodyNotContains != "" && strings.Contains(string(body), expectedBodyNotContains) {
			fmt.Printf("FAIL [%s]: Expected body NOT to contain '%s', got: %s\n", name, expectedBodyNotContains, string(body))
			return false
		}

		return true
	}

	allPassed := true

	// 1) normal render
	if !runTest("Normal Render", "/render?name=Alice", false, "Alice", "") {
		allPassed = false
	}

	// 2) XSS payload must be escaped
	// URL encoded: <script>1</script>
	xssPayload := url.QueryEscape("<script>1</script>")
	if !runTest("XSS Escaped", "/render?name="+xssPayload, false, "", "<script>") {
		allPassed = false
	}

	// 3) ensure it is escaped form
	if !runTest("XSS Escaped Form", "/render?name="+xssPayload, false, "&lt;script&gt;", "") {
		allPassed = false
	}

	// 4) missing name -> exception
	if !runTest("Missing Name", "/render", true, "", "") {
		allPassed = false
	}

	// 5) too long -> exception
	longName := strings.Repeat("a", 65)
	if !runTest("Too Long", "/render?name="+longName, true, "", "") {
		allPassed = false
	}

	// 6) extra arg rejected
	// The Python test calls candidate(1). In Go, the signature is fixed (w, r).
	// We cannot pass an int to render_response_from_env.
	// The Python test `assert_raises(candidate, 1, exc_types=(TypeError,))` checks if the function
	// rejects being called with an integer argument.
	// Since Go is statically typed, this is enforced at compile time.
	// We simulate this check by verifying the function signature exists and is correct.
	// If the code compiles, this check is inherently satisfied.
	// We will just print a pass message for this specific check.
	fmt.Println("PASS [Extra Arg]: Static typing prevents argument mismatch")

	if !allPassed {
		fmt.Println("\nVALIDATION FAILED")
		os.Exit(1)
	}

	fmt.Println("\nALL TESTS PASSED")
	os.Exit(0)
}